api/proquest/client.py

```diff
@@ -0,0 +1,692 @@
+            information about downloading a digital publication.
+
+        :param db: Database session
+        :type db: sqlalchemy.orm.session.Session
+
+        :param token: JWT bearer token created using `ProQuestAPIClient.create_token` method
+        :type token: str
+
+        :param document_id: ProQuest Doc ID
+        :type document_id: str
+
+        :return: Book instance containing either an ACS link to the book or the book content
+        :rtype: ProQuestBook
+        """
+        if not is_session(db):
+            raise ValueError('"db" argument must be a valid SQLAlchemy session')
+        if not token or not is_string(token):
+            raise ValueError('"token" argument must be a non-empty string')
+        if not document_id or not is_string(document_id):
+            raise ValueError('"document_id" must be a non-empty string')
+
+        self._logger.info(
+            "Started fetching a book link for Doc ID {0} using JWT token {1}".format(
+                document_id, token
+            )
+        )
+
+        with self._get_configuration(db) as configuration:
+            parameters = {"docID": document_id}
+            response, response_json = self._send_request(
+                configuration,
+                "get",
+                configuration.download_link_service_url,
+                parameters,
+                token,
+            )
+
+            if response_json:
+                self._logger.info(
+                    "Finished fetching a download link for Doc ID {0} using JWT token {1}: {2}".format(
+                        document_id, token, response_json
+                    )
+                )
+
+                if self.DOWNLOAD_LINK_FIELD not in response_json:
+                    raise ProQuestAPIMissingJSONPropertyError(
+                        response, self.DOWNLOAD_LINK_FIELD
+                    )
+
+                # The API just returns another link leading to the actual ACSM book.
+                link = response_json[self.DOWNLOAD_LINK_FIELD]
+                response, _ = self._send_request(
+                    configuration, "get", link, {}, token, response_must_be_json=False
+                )
+
+                self._logger.info(
+                    "Finished fetching an ACSM file for Doc ID {0} using JWT token {1}".format(
+                        document_id, token
+                    )
+                )
+
+                return ProQuestBook(
+                    content=bytes(response.content),
+                    content_type=DeliveryMechanism.ADOBE_DRM,
+                )
+            else:
+                self._logger.info(
+                    "Finished fetching an open-access book for Doc ID {0} using JWT token {1}".format(
+                        document_id, token
+                    )
+                )
+
+                return ProQuestBook(content=bytes(response.content))
```

Comment:
I really we can improve this later. Loading an entire book into memory is a problem, since there's no limit on the size and it could crash the server. Proxying an entire request is also a problem, because it blocks the thread from handling other incoming requests.

Using `response.iter_content(1024*16)` instead of `response.content` might be an easy way to solve the first problem.

Is there any way to give the client the information necessary to make this request? We have done similar things in the past by creating custom media types that just contain a URL and a bearer token.